odoo/addons/base/models/res_currency.py

```diff
@@ -248,26 +281,52 @@ class CurrencyRate(models.Model):
         ('currency_rate_check', 'CHECK (rate>0)', 'The currency rate must be strictly positive.'),
     ]
 
+    def _raise_different_one(self):
+        raise UserError(_("A company can not have a rate different than 1 for its own currency."))
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        self.env['res.company'].browse({
+            vals['company_id'] for vals in vals_list if 'company_id' in vals
+        })._normalize_rate()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        self.company_id._normalize_rate()
+        return res
+
+    def unlink(self):
+        company = self.company_id
+        res = super().unlink()
+        company._normalize_rate()
+        return res
+
     def _get_latest_rate(self):
-        return self.search([
-            ('currency_id', '=', self.currency_id.id),
-            ('company_id', '=', self.company_id.id or self.env.company.id),
-            ('name', '<=', self.name or fields.Date.today()),
-        ], order="name desc", limit=1)
+        rates = self.currency_id.rate_ids.filtered(
+            lambda x: x.rate and x.company_id == self.company_id and x.name < self.name).sorted('name')
+        return rates and rates[-1] or rates
+
+    @api.onchange('inverse_rate')
+    def _inverse_inverse_rate(self):
+        for currency_rate in self:
+            currency_rate.rate = 1.0 / currency_rate.inverse_rate
 
-    @api.depends('currency_id', 'company_id', 'name')
+    @api.depends('currency_id', 'company_id')
```

Comment:
Why are you removing `name` from the dependence? It is used in `_get_latest_rate`